_studio/shared/umc/codec/mpeg2_dec/src/umc_mpeg2_frame.cpp

```diff
@@ -0,0 +1,316 @@
+// Copyright (c) 2018 Intel Corporation
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include "umc_defs.h"
+#if defined (MFX_ENABLE_MPEG2_VIDEO_DECODE)
+
+#include "umc_mpeg2_defs.h"
+#include "umc_mpeg2_slice.h"
+#include "umc_mpeg2_frame.h"
+
+namespace UMC_MPEG2_DECODER
+{
+    MPEG2DecoderFrameInfo::MPEG2DecoderFrameInfo(MPEG2DecoderFrame& Frame)
+        : frame(Frame)
+        , isField(false)
+        , isBottomField(false)
+        , frameType((FrameType)0)
+        , isFilled(false)
+        , sliceCount(0)
+    {
+    }
+
+    MPEG2DecoderFrameInfo::~MPEG2DecoderFrameInfo()
+    {
+        std::for_each(slices.begin(), slices.end(),
+            std::default_delete<MPEG2Slice>()
+        );
+    }
+
+    void MPEG2DecoderFrameInfo::Reset()
+    {
+        isField       = false;
+        isBottomField = false;
+        frameType     = (FrameType)0;
+        isFilled      = false;
+
+        std::for_each(slices.begin(), slices.end(),
+            std::default_delete<MPEG2Slice>()
+        );
+
+        slices.resize(0);
+        sliceCount = 0;
+        FreeReferenceFrames();
+
+    }
+
+    void MPEG2DecoderFrameInfo::AddSlice(MPEG2Slice* slice)
+    {
+        if (0 == sliceCount) // on the first slice
+        {
+            const auto pic    = slice->GetPicHeader();
+            const auto picExt = slice->GetPicExtHeader();
```

Comment:
auto doesn't deduce reference, so here most probably copy occurs. In almost all cases it is faster than indirection caused by usage of reference, so I won't treat it as problem